lib/srv/db/sqlserver/kinit/kinit.go

```diff
@@ -0,0 +1,314 @@
+	RealmName string
+	// KDCHostName is the key distribution center host name (usually AD host, like ad.example.com)
+	KDCHostName string
+	// AdminServerName is the admin server name (usually AD host)
+	AdminServerName string
+
+	dataDir   string
+	userName  string
+	cacheName string
+
+	certPath string
+	keyPath  string
+	binary   string
+
+	command    commandGenerator
+	certGetter CertGetter
+
+	ldapCertificate *x509.Certificate
+	log             logrus.FieldLogger
+}
+
+// CertGetter is an interface for getting a new cert/key pair along with a CA cert
+type CertGetter interface {
+	// GetCertificateBytes returns a new cert/key pair along with a CA for use with x509 Auth
+	GetCertificateBytes(ctx context.Context) (certPEM, keyPEM, caCert []byte, err error)
+}
+
+// DBCertGetter obtains a new cert/key pair along with the Teleport database CA
+type DBCertGetter struct {
+	// Auth is the auth client
+	Auth auth.ClientI
+	// KDCHostName is the name of the key distribution center host
+	KDCHostName string
+	// RealmName is the kerberos realm name (domain name)
+	RealmName string
+	// AdminServerName is the name of the admin server. Usually same as the KDC
+	AdminServerName string
+	// UserName is the database username
+	UserName string
+	// LDAPCA is the windows ldap certificate
+	LDAPCA *x509.Certificate
+}
+
+// GetCertificateBytes returns a new cert/key pem and the DB CA bytes
+func (d *DBCertGetter) GetCertificateBytes(ctx context.Context) (certPEM, keyPEM, caCert []byte, err error) {
+	clusterName, err := d.Auth.GetClusterName()
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+
+	certPEM, keyPEM, err = windows.CertKeyPEM(ctx, d.UserName, d.RealmName, time.Second*60*60, clusterName.GetClusterName(), windows.LDAPConfig{
+		Addr:               d.KDCHostName,
+		Domain:             d.RealmName,
+		Username:           d.UserName,
+		InsecureSkipVerify: false,
+		ServerName:         d.AdminServerName,
+		CA:                 d.LDAPCA,
+	}, d.Auth)
+
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+
+	dbCA, err := d.Auth.GetCertAuthority(ctx, types.CertAuthID{
+		Type:       types.DatabaseCA,
+		DomainName: clusterName.GetClusterName(),
+	}, true)
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+
+	keyPairs := dbCA.GetActiveKeys().TLS
+	for _, keyPair := range keyPairs {
+		if keyPair.KeyType == types.PrivateKeyType_RAW {
+			caCert = keyPair.Cert
+		}
+	}
+
+	if caCert == nil {
+		return nil, nil, nil, trace.Wrap(errors.New("no certificate authority was found in userCA active keys"))
+	}
+
+	return
+}
+
+// UseOrCreateCredentials uses an existing cache or creates a new one
+func (k *CommandLineInitializer) UseOrCreateCredentials(ctx context.Context) (*credentials.CCache, error) {
+	tmp, err := os.MkdirTemp("", "kinit")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	defer func() {
+		err = os.RemoveAll(tmp)
+		if err != nil {
+			k.log.Error(err)
```

Comment:
We should try and add more context to this error.

```suggestion
			k.log.WithError(err).Error("Failed to clear up kinit temporary directory)
```